File: weeks/3/3/solution.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

typedef long long ll;

constexpr int MAX_NUM_ISLANDS = int(1e3+1);

vector<int> bridges[MAX_NUM_ISLANDS];
bool visited[MAX_NUM_ISLANDS];

void walk(int distanceLimit, int destIsland) {
  queue<pair<int, int>> bfsQueue;
  bfsQueue.push({1, 0}); // island, distance

  while (!bfsQueue.empty()) {
    auto [island, distance] = bfsQueue.front();
    bfsQueue.pop();

    // 주어진 거리를 넘기면 스킵
    if (distance >= distanceLimit) {
      continue;
    }

    // 방문한 적 없는 다음 섬들을 큐에 대기
    for (int nextIsland : bridges[island]) {
      // 목적지 섬에 도착하면 가능하다고 리턴
      if (nextIsland == destIsland) {
        cout << "YES" << '\n';
        return;
      }

      if (visited[nextIsland]) {
        continue;
      }
      visited[nextIsland] = true;

      bfsQueue.push({ nextIsland, distance + 1});
    }
  }

  // 목적지 섬에 도달할 수 없으면 불가능하다고 리턴
  cout << "NO" << '\n';
}

void solve() {
  int numIslands, numBridges, distanceLimit;
  cin >> numIslands >> numBridges >> distanceLimit;

  for (int i = 0; i < numBridges; ++i) {
    int island1, island2;
    cin >> island1 >> island2;
    bridges[island1].push_back(island2);
    bridges[island2].push_back(island1);
  }

  int destIsland = numIslands;

  visited[1] = true;

  walk(distanceLimit, destIsland);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  solve();
}

File: weeks/3/3/solution.py
import sys
sys.setrecursionlimit(10**6)
readline = sys.stdin.readline
from collections import deque

MAX_NUM_ISLANDS = int(1e3+1)

numIslands, numBridges, distanceLimit = map(int, readline().split())

bridges = [[] for _ in range(MAX_NUM_ISLANDS)]
for _ in range(numBridges):
    island1, island2 = map(int, readline().split())
    bridges[island1].append(island2)
    bridges[island2].append(island1)

destIsland = numIslands

visited = [False] * numIslands
visited[1] = True

def walk():
    bfsQueue = deque()
    bfsQueue.append((1, 0))

    while len(bfsQueue) > 0:
        island, distance = bfsQueue.popleft()

        # 주어진 거리를 넘기면 스킵
        if distance >= distanceLimit:
            continue

        # 방문한 적 없는 다음 섬들을 큐에 대기
        for nextIsland in bridges[island]:
            # 목적지 섬에 도착하면 가능하다고 리턴
            if nextIsland == destIsland:
                return "YES"

            if visited[nextIsland]:
                continue
            visited[nextIsland] = True

            bfsQueue.append((nextIsland, distance+1))

    # 목적지 섬에 도달할 수 없으면 불가능하다고 리턴
    return "NO"

print(walk())
